### DISCUSSION TITLE
How do get Telemetry Database data?

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#775] [Q&A]
Hi,

I am saving telemetry data in my components and I can see them on GDS but GDS is only for debugging. How can I get all the data inside Telemetry Database in a text or csv format (or any other useful format) to downlink? Can I get all the data or do I have iterate through their ids? Can I delete the data because I don't want any telemetry older than x days?

If not what are some recommendations?

Thanks

**Comment by Joshua-Anderson** [comment]
The F' GDS is really meant for development and testing, and because of that it doesn't have any sort of telemetry / event persistence system. It'll show you live data and that's about it.

Generally in real operations you want to save all the data you can get from your spacecraft.

Personally I recommend building a tool that can store your received telemetry into a database (I don't think anybody's built anu open source F' telemetry to database tools yet). Once you have your telemetry in a database, it becomes much easier to visualize and query your telemetry.

If you want to store telemetry on the spacecraft side I recommend taking a look at the ComLogger component. It can log telemetry channels and events to disk, which can then be downlinked at a later time.

**Reply by sheershak-agarwal** [reply]
Querying on satellite is unnecessary so I am not keen on designing a database for it.
But I have a few question about COMLogger, because it might be the tool for me.
So I can create/open files using COMLogger and update data onto those files through comIn port. Then I can close the file.

Q1. Can I create multiple files? and write to them? If yes then how?
Q2. Is the file created accessible by any component or how do access the file to downlink?
Q3. What will be the format of the file?
Q4. Can I write in the file after closing it by opening it back again? or closing means deleting the file?
Q5. How can I delete the file from disk?

**Reply by Joshua-Anderson** [reply]
A lot of these answers are up to you! The comlogger in Svc/ComLogger I believe opens a file with the name of the current date and automatically rotates when the filesize gets too large.

> Q1. Can I create multiple files? and write to them? If yes then how?

You can multiple comlogger instantiations, and have different components log to different comlogger files.

> Q2. Is the file created accessible by any component or how do access the file to downlink?

It'll just be a normal file on the filesystem. Usually you'd use something like filedownlink to downlink the comlogger file.

> Q3. What will be the format of the file?

It'll log data in the standard F' serialized format, the same way data is serialized when transmitted over the radio.
I don't think there's currently any generic tools for parsing comlogger files, but it's pretty similar to what the GDS does so one could probably be built using the GDS parsing code.

> Q4. Can I write in the file after closing it by opening it back again? or closing means deleting the file?

You can tweak the comlogger service to do whatever you want. Right now if it doesn't have a open file it'll open one with the current date, then rotate to prevent individual files from getting too large. 

> Q5. How can I delete the file from disk?

You can delete files using whatever technique you plan to use to manage the filesystem. the FileManager component can delete files.

**Reply by sheershak-agarwal** [reply]
Thank you, can you please elaborate on what does rotate mean in "opens a file with the name of the current date and automatically **rotates** when the filesize gets too large"?

**Reply by Joshua-Anderson** [reply]
ComLogger does a kind of log rotation... when a file goes over a specific limit it'll close the existing log file and open a new log file using the same naming scheme. This breaks the log into files of manageable chunks.
